genienlp/models/general_seq2seq.py

```diff
@@ -126,7 +145,51 @@ def get_encoder_loss(self, context_rnn_state):
             
         return encoder_loss
         
-        
+    def get_output_embeddings(self):
+        return self.decoder.decoder_embeddings
+
+    def prepare_inputs_for_generation(self, input_ids, past, attention_mask, use_cache, batch, generation_dict):
+        expansion_factor = input_ids.shape[0] // len(batch.example_id)
+        return {"batch": batch, "past": past, "current_token_id": input_ids[:,-1:], "expansion_factor": expansion_factor, "generation_dict": generation_dict}
+
+    def _reorder_cache(self, past, beam_idx):
+        past.reorder(beam_idx)
+        return past
+
+    def generate(self,
+                 batch,
+                 max_output_length,
+                 num_outputs,
+                 temperature,
+                 repetition_penalty,
+                 top_k,
+                 top_p,
+                 num_beams,
+                 no_repeat_ngram_size,
+                 do_sample
+                 ):
+        self.config.vocab_size = len(batch.decoder_vocab)
+        batch_size = len(batch.example_id)
+        input_ids = torch.full((batch_size, 1), self.decoder.init_idx, dtype=torch.long, device=batch.context.value.device)
+
+        generated = super().generate(input_ids=input_ids,
+                                     batch=batch,
+                                     max_length=max_output_length,
+                                     bos_token_id=self.decoder.init_idx,
+                                     pad_token_id=batch.decoder_vocab.pad_idx,
+                                     num_return_sequences=num_outputs,
+                                     repetition_penalty=repetition_penalty,
+                                     temperature=temperature,
+                                     eos_token_id=batch.decoder_vocab.eos_idx,
+                                     top_k=top_k,
+                                     top_p=top_p,
+                                     num_beams=num_beams,
+                                     no_repeat_ngram_size=no_repeat_ngram_size,
+                                     do_sample=do_sample,
+                                     generation_dict={'max_output_length': max_output_length, 'num_beams': num_beams}
+                                    )
+        generated = generated[:, 1:].cpu().apply_(self.decoder.map_to_full).to(batch.context.value.device) # remove bos token and map to full vocabulary
```

Comment:
Why do you remove the first token? The first token of the output is _not_ the init token, and you normally wouldn't remove it...

Reply:
The new generation code (from the `transformers` package) outputs BOS token as well.

Reply:
I think it's better that special tokens will be handled inside numericalizer.reverse(.) which is the case already, and return full "generated" here.